### PR TITLE
Handle reference issue when multiple conformation provided in `[caprieval]`

### PR DESCRIPTION
**Pull request by VGPReys**
## Checklist

- [x] Tests added for the new code
- [ ] Documentation added for the code changes
- [x] Does not break licensing
- [x] Does not add any dependencies, if it does please add a thorough explanation

## Summary of the Pull Request  

Add machinery to handle multiple conformation in input `reference_fname` in `[caprieval]` module.

When multiple conformations are provided:
- the `capri_ss.tsv` file will contain an additional column `ref_id` allowing to traceback what was the best reference for each complexes
- an additional file `capri_ss_multiref.tsv` will be created: it contains all the performance metrics against all references. Therefore each input complex will have X number of entries (X being the number of reference conformations).

## Related Issue

#1132

### CODE REVIEW

File: src/haddock/modules/analysis/caprieval/__init__.py [hunk 1]
```diff
@@ -69,6 +73,92 @@ def is_nested(models: list[Union[PDBFile, list[PDBFile]]]) -> bool:
...
+        wc_return = sys.stdout.read()  # Read output
+        # Restore original stdout
+        sys.stdout.close()
+        sys.stdout = original_stdout
+        # Parse output
+        for line in wc_return.split("\n"):
+            if "No. models" in line:
+                sline = line.strip().split()
+                nb_models = int(sline[-1])
+                break
```
Comment (mgiulini) [inline src/haddock/modules/analysis/caprieval/__init__.py:113]: extremely complex indeed...

**Review by mgiulini — APPROVED** [review]
quite complicated syntax but I guess unavoidable..let's go for it

File: src/haddock/modules/analysis/caprieval/__init__.py (hunk 1 above)
Comment (VGPReys) [reply src/haddock/modules/analysis/caprieval/__init__.py:113]: yes, the pdb_tools `pdb_wc` is not returning anything... :'(

**Comment by VGPReys** [comment]
About to create a new PR to be merged to this one, expending the approach to evaluate on multiple references instead of only selecting the first.

File: src/haddock/modules/analysis/caprieval/__init__.py [hunk 2]
```diff
@@ -69,6 +73,92 @@ def is_nested(models: list[Union[PDBFile, list[PDBFile]]]) -> bool:
...
+        with open(reference, "r") as fh:
+            pdb_wc(fh, "m")
+        # Get output
+        sys.stdout.seek(0)  # Jump to the start
+        wc_return = sys.stdout.read()  # Read output
+        # Restore original stdout
+        sys.stdout.close()
+        sys.stdout = original_stdout
+        # Parse output
+        for line in wc_return.split("\n"):
```
Comment (VGPReys) [inline src/haddock/modules/analysis/caprieval/__init__.py:109]:
```suggestion
        # using here `\n` (and not os.linesep) as it is the output for pdb_wc
        for line in wc_return.split("\n"):
```

**Comment by mgiulini** [comment]
> About to create a new PR to be merged to this one, expending the approach to evaluate on multiple references instead of only selecting the first.

why? didn't we agree that evaluating on multiple references would be confusing?

**Comment by VGPReys** [comment]
> > About to create a new PR to be merged to this one, expending the approach to evaluate on multiple references instead of only selecting the first.
> 
> why? didn't we agree that evaluating on multiple references would be confusing?

Yes, but in the end I think I found an ok way to handle this without a very minimum number of modifications.
It also expends the possibilities of the module, which I feel is nice.
Because splitting the references into X files and running X `[caprieval]` and having to parse X files to gather the data seems (to me) difficult for unprepared users.
Of course, they can still do it if they want.

**Comment by mgiulini** [comment]
I tried this on an independent example:
I see that capri calculations are performed, but I cannot see where the CAPRI results for the various "non-best" models are saved. If we calculate these numbers then we have to provide them to the users somehow.

**Comment by VGPReys** [comment]
> I tried this on an independent example: I see that capri calculations are performed, but I cannot see where the CAPRI results for the various "non-best" models are saved. If we calculate these numbers then we have to provide them to the users somehow.

To not disturb the user, the outputs from `caprieval`  are the same.
The only difference (introduced by https://github.com/haddocking/haddock3/pull/1136) is the fact that reported values only corresponds to the best reference, which may be different for each model.

Do you think that additionally we should create a `ref_X_capri_ss.tsv` for each `X` reference conformation ?

**Comment by mgiulini** [comment]
> > I tried this on an independent example: I see that capri calculations are performed, but I cannot see where the CAPRI results for the various "non-best" models are saved. If we calculate these numbers then we have to provide them to the users somehow.
> 
> To not disturb the user, the outputs from `caprieval` are the same. The only difference (introduced by #1136) is the fact that reported values only corresponds to the best reference, which may be different for each model.
> 
> Do you think that additionally we should create a `ref_X_capri_ss.tsv` for each `X` reference conformation ?

If this is the desired logic, I think so. Otherwise it would be impossible to trace back which numbers correspond to which reference, correct?

**Comment by VGPReys** [comment]
> > > I tried this on an independent example: I see that capri calculations are performed, but I cannot see where the CAPRI results for the various "non-best" models are saved. If we calculate these numbers then we have to provide them to the users somehow.
> > 
> > 
> > To not disturb the user, the outputs from `caprieval` are the same. The only difference (introduced by #1136) is the fact that reported values only corresponds to the best reference, which may be different for each model.
> > Do you think that additionally we should create a `ref_X_capri_ss.tsv` for each `X` reference conformation ?
> 
> If this is the desired logic, I think so. Otherwise it would be impossible to trace back which numbers correspond to which reference, correct?

Indeed, for now it is not possible to trace back which numbers corresponds to which reference.
But this is intended to behave that way, as the question solved is: what is the quality of my docking model against any of the reference conformations (mostly used for NMR ensemble).

To solve the traceback issue, I see multiple solutions:
- add the machinery to generate multiple capri_ss / capri_clt for each reference conformation.
- add a `ref_id` column in the capri_ss, referencing to the index of the best reference in the conformational ensemble for each input model
- do both

**Comment by amjjbonvin** [comment]
I would vote for for adding a ref_id column.  

> 
> add a ref_id column in the capri_ss, referencing to the index of the best reference in the conformational ensemble for each input model

**Comment by amjjbonvin** [comment]
Make sure to check this is not breaking any of the analysis part, nor the haddock3-webapp display of results.

**Comment by VGPReys** [comment]
> Make sure to check this is not breaking any of the analysis part, nor the haddock3-webapp display of results.

Very good point !

Should I also update the analysis scripts to display the `ref_id` data in the table ?

**Comment by amjjbonvin** [comment]
Not needed - this only makes sense in a benchmarking scenario - in real applications you won’t know the reference…
We could for simplicity even think of writing a separate file that contains ref_id (only in case where an ensemble is provided)


> Should I also update the analysis scripts to display the ref_id data in the table ?
>

**Comment by VGPReys** [comment]
Analysis plots are rendering without issues.

Also, now, when multiple conformations are provided:
- the `capri_ss.tsv` file will contain an additional column `ref_id` allowing to traceback what was the best reference for each complexes
- an additional file `capri_ss_multiref.tsv` will be created: it contains all the performance metrics against all references, and the `ref_id` column allowing for traceback. Therefore each input complex will have X number of entries (X being the number of conformations in the reference file).

**Review by mgiulini — APPROVED** [review]
working as expected!

**Pull request merged**